File: src/statesman/commands/Command.java
package statesman.commands;

import java.util.HashMap;

import statesman.GameException;

public abstract class Command {

    private static HashMap<String, Command> _commands;

    public static HashMap<String, Command> getCommands() {
        if (_commands == null) {
            _commands = new HashMap<>();
            _commands.put(PrintCommand.ID_PRINT, PrintCommand.getDefault());
            _commands.put(PrintCommand.ID_PRINTC, PrintCommand.getDefault());
            _commands.put(PrintCommand.ID_PRINTR, PrintCommand.getDefault());
            _commands.put(SceneCommand.ID_SCENE, SceneCommand.getDefault());
            _commands.put(CallCommand.ID_CALL, CallCommand.getDefault());
            _commands.put(CallCommand.ID_GOTO, CallCommand.getDefault());
            _commands.put(CallCommand.ID_CALL_GLOB, CallCommand.getDefault());
            _commands.put(CallCommand.ID_SUPER, CallCommand.getDefault());
            _commands.put(JumpCommand.ID_JMP, JumpCommand.getDefault());
            _commands.put(JumpCommand.ID_IJMP, JumpCommand.getDefault());
            _commands.put(JumpCommand.ID_SJMP, JumpCommand.getDefault());
            _commands.put(JumpCommand.ID_RET, JumpCommand.getDefault());
            _commands.put(SwitchSetCommand.ID_SET,
                    SwitchSetCommand.getDefault());
            _commands.put(InventoryCommand.ID_INV,
                    InventoryCommand.getDefault());
            _commands.put(PointsCommand.ID_POINTS, PointsCommand.getDefault());
            _commands.put(MenuCommand.ID_SAVE, MenuCommand.getDefault());
            _commands.put(MenuCommand.ID_LOAD, MenuCommand.getDefault());
            _commands.put(MenuCommand.ID_QUIT, MenuCommand.getDefault());
        }
        return _commands;
    }

    public static Command find(String[] arguments) throws GameException {
        String commandId = arguments[0].toLowerCase();
        Command command = getCommands().get(commandId);
        if (command == null) {
            return null;
        }
        return command.fromText(commandId, arguments);
    }

    public abstract Command fromText(String id, String[] arguments)
            throws GameException;

    public abstract void execute();

}

File: src/Commands/Command.cs
using System.Reflection;

namespace Statesman.Commands
{
    public abstract class Command
    {
        public const string kGeneratorMethodName = "FromText";

        public static Dictionary<string, Type> All { get; private set; }

        static Command()
        {
            All = new Dictionary<string, Type>
            {
                // XXX: manually input IDs of new commands here!
                { PrintCommand.kIdPrintSingle, typeof(PrintCommand) },
                { PrintCommand.kIdPrintConcatenate, typeof(PrintCommand) },
                { PrintCommand.kIdPrintRandom, typeof(PrintCommand) },
                { JumpCommand.kIdJump, typeof(JumpCommand) },
                { JumpCommand.kIdSwitchJump, typeof(JumpCommand) },
                { JumpCommand.kIdInventoryJump, typeof(JumpCommand) },
                { JumpCommand.kIdReturn, typeof(JumpCommand) },
                { SwitchSetCommand.kIdSwitchSet, typeof(SwitchSetCommand) },
                { ConditionalCommand.kIdConditional, typeof(ConditionalCommand) },
                { InventoryCommand.kIdInventory, typeof(InventoryCommand) },
                { SceneCommand.kIdScene, typeof(SceneCommand) },
                { CallCommand.kIdCall, typeof(CallCommand) },
                { CallCommand.kIdGoto, typeof(CallCommand) },
                { CallCommand.kIdCallGlobal, typeof(CallCommand) },
                { CallCommand.kIdSuper, typeof(CallCommand) },
                { MenuCommand.kIdQuit, typeof(MenuCommand) },
                { MenuCommand.kIdSave, typeof(MenuCommand) },
                { MenuCommand.kIdLoad, typeof(MenuCommand) },
                { PointsCommand.kIdPoints, typeof(PointsCommand) },
            };
        }

        public static Command Find(string[] arguments)
        {
            string commandId = arguments[0].ToLowerInvariant();
            if (All.TryGetValue(commandId, out Type commandType))
            {
                MethodInfo generatorMethod =
                    commandType.GetMethod(kGeneratorMethodName);
                if (generatorMethod == null)
                {
                    throw new InvalidOperationException("Missing generator method.");
                }
                Command command = (Command)generatorMethod
                    .Invoke(null, new object[] { commandId, arguments });
                return command;
            }
            return null;
        }

        public virtual void Execute() { }
    }
}
